File: algorithms/medianOfTwoSortedArrays/medianOfTwoSortedArrays.cpp
// Source : https://oj.leetcode.com/problems/median-of-two-sorted-arrays/
// Author : Hao Chen
// Date   : 2014-07-22

/********************************************************************************** 
* 
* There are two sorted arrays A and B of size m and n respectively. 
* Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
*               
**********************************************************************************/

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        size_t len = nums1.size() + nums2.size();
        return (len % 2 == 1) ?
            findKthSmall (nums1, nums2, len / 2 + 1) :
            (findKthSmall (nums1, nums2, len / 2 + 1) + 
            findKthSmall (nums1, nums2, len / 2)) / 2;
    }
private:
    double findKthSmall (const vector<int> &nums1, const vector<int> &nums2, size_t k) {
        size_t len1 = nums1.size(), len2 = nums2.size();
        if (len1 > len2) // ensure that len1 is always not greater than len2
            return findKthSmall(nums2, nums1, k);
        // simple situations
        if (len1 == 0)
            return nums2[k - 1];
        if (k == 1)
            return min(nums1[0], nums2[0]);
        size_t pNums1 = min(k / 2, len1), pNums2 = k - pNums1; // two working pointers (size_t)
        // algorithm: divide and conquer, binary search
        if (nums1[pNums1 - 1] < nums2[pNums2 - 1]) {
            // this means the former pNums1 elements in nums1 are too small
            // we could drop them and findKthSmall in temp and nums2, and k is k - pNums1
            vector<int> temp((nums1.begin() + pNums1), nums1.end());
            return findKthSmall(temp, nums2, k - pNums1);
        }
        else if (nums1[pNums1 - 1] > nums2[pNums2 - 1]) {
            // this means the former pNums2 elements in nums2 are too small
            // similar method compared to the first case
            vector<int> temp((nums2.begin() + pNums2), nums2.end());
            return findKthSmall(nums1, temp, k - pNums2);
        }
        else
            // this means we've found the answer, return level by level
            return nums1[pNums1 - 1];
    }
};

File: algorithms/java/src/MedianOfTwoSortedArrays/MedianOfTwoSortedArrays.java
// Source : https://leetcode.com/problems/median-of-two-sorted-arrays/
// Author : Diego Ruiz Piqueras (Pikeras72)
// Date   : 23-04-2022

/***************************************************************************************************** 
 *
 * Given two sorted arrays nums1 and nums2 of size m and n respectively,
 * return the median of the two sorted arrays.
 * 
 * The overall run time complexity should be O(log (m+n)).
 * 
 * Example 1:
 * 
 * Input: nums1 = [1,3], nums2 = [2]
 * Output: 2.00000
 * Explanation: merged array = [1,2,3] and median is 2.
 * 
 * Example 2:
 *
 * Input: nums1 = [1,2], nums2 = [3,4]
 * Output: 2.50000
 * Explanation: merged array = [1,2,3,4] and median is (2 + 3) / 2 = 2.5.
 * 
 * 
 * Constraints:
 * 
 *  nums1.length == m
 *  nums2.length == n
 *  0 <= m <= 1000
 *  0 <= n <= 1000
 *  1 <= m + n <= 2000
 * -10^6 <= nums1[i], nums2[i] <= 10^6
 * 
 * Explanation of the solution:
 * 
 * 1. Concatenate both arrays (int[] concatenate).
 * 
 * 2. Sort the 'concatenate' array.
 * 
 * 3. If 'concatenate' have an odd nuber of elements return the element in the middle, otherwise return the sum of the middle numbers.
 ******************************************************************************************************/

class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        double c;
        int[] concatenate = new int[nums1.length + nums2.length];
        System.arraycopy(nums1, 0, concatenate, 0, nums1.length);              //Add the first array
        System.arraycopy(nums2, 0, concatenate, nums1.length, nums2.length);   //Add the second array
        Arrays.sort(concatenate);
        if (concatenate.length % 2 == 0){
            c = concatenate[concatenate.length/2-1]+concatenate[concatenate.length/2];
            return c/2;
        }else {
            return concatenate[concatenate.length/2];
        }
    }
}
